core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -816,8 +821,9 @@ public ShareAcquiredRecords acquire(
 
                 // Compute if the batch is a full match.
                 boolean fullMatch = checkForFullMatch(inFlightBatch, firstBatch.baseOffset(), lastOffsetToAcquire);
-
-                if (!fullMatch || inFlightBatch.offsetState() != null) {
+                long numRecordsToAcquire = inFlightBatch.lastOffset() - inFlightBatch.firstOffset() + 1;
```

Comment:
The name should be `numRecordsInBatch` as we don't know yet if all records from the batch can be acquired or not.
```suggestion
                long numRecordsInBatch = inFlightBatch.lastOffset() - inFlightBatch.firstOffset() + 1;
```